### PR TITLE
[WIP] Revisit sypgp to enable better testing coverage

### PR DESCRIPTION
**Pull request by gvallee**
**Description of the Pull Request (PR):**

Improve testing coverage for sypgp package. Addresses issue #2559

**This fixes or addresses the following GitHub issues:**

- Addresses #2559 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

**Comment by gvallee** [comment]
This is a set of modifications that improve the testing coverage for the sypgp package.

I would like to have feedback about the changes i did to the package so we can redirect stdin using a file and with that enable automatic testing. In fact, these modifications improve coverage from ~20% to ~70%.
I did quite an extensive testing on Linux and did not see any change in the behavior of the various functions of the package.

I also believe others are currently touching PGP support right now so if these modifications are of interest, i have no problem leaving this PR open until all PGP modifications are pushed in; i will update these modifications then.

File: pkg/sypgp/sypgp.go [hunk 1]
```diff
@@ -48,27 +48,64 @@ WARNING: this may overwrite a previous token if ~/.singularity/sylabs-token exis
...
+// The file can be a normal file or os.Stdin.
+// Note that we could imagine a simpler code but we want to make sure that
+// the code works properly in the normal case with the default Stdin and when
+// redirecting stdin (for testing or when using pipes).
+func askQuestionUsingGenericDescr(f *os.File) (string, error) {
+	// Get the initial position in the buffer so we can later seek the correct
+	// position based on how much data we read. Doing so, we can still benefit
+	// from buffered io and still have a fine-grain control over reading
+	// operations
+	pos, _ := f.Seek(0, os.SEEK_CUR)
```
Comment (cclerget) [inline pkg/sypgp/sypgp.go]: If `os.Stdin` is a pipe, seek will fail

File: pkg/sypgp/sypgp.go [hunk 2]
```diff
@@ -212,6 +249,10 @@ func LoadPubKeyring() (openpgp.EntityList, error) {
 
 // PrintEntity pretty prints an entity entry
 func PrintEntity(index int, e *openpgp.Entity) {
+	if e == nil {
+		return
```
Comment (cclerget) [inline pkg/sypgp/sypgp.go:259]: maybe display a warning as we don't return any error here

File: pkg/sypgp/sypgp.go [hunk 3]
```diff
@@ -255,6 +296,10 @@ func PrintPrivKeyring() (err error) {
 
 // StorePrivKey stores a private entity list into the local key cache
 func StorePrivKey(e *openpgp.Entity) (err error) {
+	if e == nil {
+		return
```
Comment (cclerget) [inline pkg/sypgp/sypgp.go]: safest to return an explicit error message

File: pkg/sypgp/sypgp.go [hunk 4]
```diff
@@ -269,6 +314,10 @@ func StorePrivKey(e *openpgp.Entity) (err error) {
 
 // StorePubKey stores a public key entity list into the local key cache
 func StorePubKey(e *openpgp.Entity) (err error) {
+	if e == nil {
+		return
```
Comment (cclerget) [inline pkg/sypgp/sypgp.go]: same here

File: pkg/sypgp/sypgp.go [hunk 5]
```diff
@@ -461,6 +514,10 @@ func GenKeyPair(keyServiceURI string, authToken string) (entity *openpgp.Entity,
 
 // DecryptKey decrypts a private key provided a pass phrase
 func DecryptKey(k *openpgp.Entity) error {
+	if k == nil {
+		return nil
```
Comment (cclerget) [inline pkg/sypgp/sypgp.go]: should return an error because this is a misuse of this function

File: pkg/sypgp/sypgp.go [hunk 6]
```diff
@@ -476,11 +533,18 @@ func DecryptKey(k *openpgp.Entity) error {
 
 // EncryptKey encrypts a private key using a pass phrase
 func EncryptKey(k *openpgp.Entity, pass string) (err error) {
+	if k == nil {
+		return nil
```
Comment (cclerget) [inline pkg/sypgp/sypgp.go]: same here

**Review by cclerget — CHANGES_REQUESTED** [review]

**Comment by gvallee** [comment]
For reference, this work will create major conflicts with #3290, which should be merged first since this PR needs work to address the rebase that got ugly.

**Comment by ArangoGutierrez** [comment]
How is this PR going, still a WIP?

**Comment by gvallee** [comment]
Some work is still ongoing around GPG and signing (mainly #3300) and I do not want to interfere with that work (it is at the point the priority over testing coverage). As soon as it is completed, @mem will merge his work on sygpg.go and then I will merge my work on top of it. So it is almost there, we just need a little more time. :)

**Comment by gvallee** [comment]
PR #3705 provides a clean PR to introduce the modifications required to handle asking questions in a way that can be automated for testing. It also introduce a few new unit tests for functions that could not be easily tested in the past, based on these modifications.
Better testing coverage is still possible and coming in a separate PR.

**Comment by sashayakovtseva** [comment]
@gvallee Still a WIP or can be closed since #3705 is merged?

**Pull request closed without merging**